causalpy/transform_optimization.py

```diff
@@ -0,0 +1,602 @@
+    arima_order : tuple of (int, int, int), optional
+        ARIMA order (p, d, q) when error_model="arimax".
+
+    Returns
+    -------
+    rmse : float
+        Root mean squared error of the fit.
+    ols_result : RegressionResultsWrapper
+        Fitted OLS or ARIMAX model object.
+    """
+    # Build baseline design matrix
+    X_baseline = np.asarray(dmatrix(base_formula, data))
+
+    # Apply transforms to treatment variable
+    x_raw = data[treatment_name].values
+    x_transformed = x_raw
+
+    if saturation is not None:
+        x_transformed = saturation.apply(x_transformed)
+    if adstock is not None:
+        x_transformed = adstock.apply(x_transformed)
+    if lag is not None:
+        x_transformed = lag.apply(x_transformed)
+
+    # Build full design matrix
+    X_full = np.column_stack([X_baseline, x_transformed])
+
+    # Get outcome
+    y = data[y_column].values
+
+    # Fit model with chosen error structure
+    if error_model == "hac":
+        # Fit OLS with HAC standard errors
+        if hac_maxlags is None:
+            n = len(y)
+            hac_maxlags = int(np.floor(4 * (n / 100) ** (2 / 9)))
+
+        ols_result = sm.OLS(y, X_full).fit(
+            cov_type="HAC", cov_kwds={"maxlags": hac_maxlags}
+        )
+    elif error_model == "arimax":
+        # Fit ARIMAX model
+        import warnings
+
+        from statsmodels.tsa.statespace.sarimax import SARIMAX
+
+        # ARIMAX requires at least as many observations as parameters
+        # Quick validation
+        n_obs = len(y)
+        n_params = X_full.shape[1] + sum(arima_order)  # exog params + ARIMA params
+        if n_obs < n_params + 10:  # Need some degrees of freedom
+            raise ValueError(
+                f"ARIMAX requires more observations. Have {n_obs}, need at least {n_params + 10}"
+            )
+
+        # Suppress convergence warnings during grid search
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            arimax_model = SARIMAX(y, exog=X_full, order=arima_order)
+            ols_result = arimax_model.fit(
+                disp=0,
+                maxiter=200,
+                method="lbfgs",
+            )
+    else:
+        raise ValueError(f"error_model must be 'hac' or 'arimax', got '{error_model}'")
+
+    # Compute RMSE
+    residuals = ols_result.resid
+    rmse = np.sqrt(np.mean(residuals**2))
+
+    return rmse, ols_result
+
+
+def estimate_transform_params_grid(
+    data: pd.DataFrame,
+    y_column: str,
+    treatment_name: str,
+    base_formula: str,
+    saturation_type: Optional[str],
+    saturation_grid: Optional[Dict[str, List[float]]],
+    adstock_grid: Optional[Dict[str, List[float]]],
+    coef_constraint: str = "nonnegative",
+    hac_maxlags: Optional[int] = None,
+    metric: str = "rmse",
+    error_model: str = "hac",
+    arima_order: Optional[Tuple[int, int, int]] = None,
+) -> Dict[str, Any]:
```

Comment:
also use `dict` as we are now over Python 3.9 (we do not need to use these upper case type hints)